### PR TITLE
bpo-45711: use exc_value instead of exc_type to determine if exc_info is valid. Add more assertions.

### PR DESCRIPTION
**Pull request by iritkatriel**
This brings us closer to removing exc_type from exc_info.

Where exc_type is used, I made it use exc_value instead and added assertions that this is ok. Then when we remove exc_type it will just be a matter of removing the assertions.

https://bugs.python.org/issue45711

### CODE REVIEW

File: Python/errors.c [hunk 1]
```diff
@@ -507,6 +526,30 @@ PyErr_SetExcInfo(PyObject *p_type, PyObject *p_value, PyObject *p_traceback)
...
+{
+    PyObject *exc_value = err_info->exc_value;
+    if (exc_value == NULL) {
+        exc_value = Py_None;
+    }
+
+    assert(exc_value == Py_None || PyExceptionInstance_Check(exc_value));
+
+    PyObject *exc_type = PyExceptionInstance_Check(exc_value) ?
+               PyExceptionInstance_Class(err_info->exc_value) :
```
Comment (gvanrossum) [inline Python/errors.c]:
```suggestion
               PyExceptionInstance_Class(err_info->exc_value) :
```
```suggestion
               PyExceptionInstance_Class(exc_value) :
```

**Review by gvanrossum — COMMENTED** [review]
I noticed that in some places you require type and value to be both "nullish" or both "not nullish" (where "nullish" means either NULL or Py_None), and in other places (in particular in `_assert_exception_type_is_redundant()`) they are required to both be the *same* nullishness (i.e., both NULL or both Py_None or neither of them nullish). Is there a reason for the difference?

File: Python/errors.c [hunk 2]
```diff
@@ -79,11 +79,15 @@ _PyErr_StackItem *
 _PyErr_GetTopmostException(PyThreadState *tstate)
 {
     _PyErr_StackItem *exc_info = tstate->exc_info;
```
Comment (gvanrossum) [inline Python/errors.c:81]: Is it worth asserting that `exc_info != NULL` here? The `while` condition assumes that and ensures that it stays that way, but the assert below the loop allows for it to be NULL. That seems inconsistent.

File: Python/errors.c [hunk 3]
```diff
@@ -471,10 +475,25 @@ _PyErr_GetExcInfo(PyThreadState *tstate,
                   PyObject **p_type, PyObject **p_value, PyObject **p_traceback)
 {
     _PyErr_StackItem *exc_info = _PyErr_GetTopmostException(tstate);
-    *p_type = exc_info->exc_type;
+
     *p_value = exc_info->exc_value;
     *p_traceback = exc_info->exc_traceback;
```
Comment (gvanrossum) [inline Python/errors.c:481]: So here we still allow exc_traceback to differ from exc_value->traceback, right? Is that going to be fixed in a separate PR?

**Comment by iritkatriel** [comment]
> I noticed that in some places you require type and value to be both "nullish" or both "not nullish" (where "nullish" means either NULL or Py_None), and in other places (in particular in `_assert_exception_type_is_redundant()`) they are required to both be the _same_ nullishness (i.e., both NULL or both Py_None or neither of them nullish). Is there a reason for the difference?


Yes. Most of these assertions are verifying that the change right next to them is valid. So if I replace `if(exc_type is nullish)` by `if(exc_value is nullish)` then I only care that their nullishness is the same.  In the case of the triplets pushed to the stack, it used to be that no-exc was pushed as "NULL, NULL, None" (None for type, NULL for value and tb).   I change that in this PR (ceval.c lines 5907-9) so now it's "NULL, None, None".  There are places where "None at the top of the stack" means there is no exc_info. So when we will remove the type we need the value to be None and not NULL, otherwise we'll get segfaults.

File: Python/errors.c [hunk 4]
```diff
@@ -471,10 +475,25 @@ _PyErr_GetExcInfo(PyThreadState *tstate,
...
+    *p_type = (*p_value != NULL && PyExceptionInstance_Check(*p_value)) ?
+              PyExceptionInstance_Class(*p_value) :
+              Py_None;
+
+    if (exc_info->exc_type == NULL || exc_info->exc_type == Py_None) {
+        assert(*p_type == Py_None);
+    }
+    else {
+        assert(*p_type == exc_info->exc_type);
+    }
```
Comment (gvanrossum) [inline Python/errors.c]: The logic here is a bit hard to follow. Let me see if this summarizes it:
```py
typ, value, traceback = exc_info
assert nullish(value) or isinstance(value, BaseException)
typ = None if isinstance(value, BaseException) else type(value)
if nullish(exc_info[0]):
    assert typ is None
else:
    assert typ is exc_info[0]
return typ, value, traceback
```
Maybe it would be easier to follow if you rewrote it like this?
```py
typ, value, traceback = exc_info
if nullish(value):
    assert nullish(type)
    typ = None
elif isinstance(value, BaseException):
    assert typ is type(value)
    typ = type(value)  # In case assertions are off and it was a lie
else:
    assert False
return typ, value, traceback
```
Which avoids referring back to `exc_info` once we've extracted everything from it. (I think the reason I found it hard to follow is that it refers back to `exc_info`.)

Also note this effectively checks
```py
assert nullish(type) == nullish(value)
```
Which is an invariant you also assert on lines 614-615 (in `_PyErr_ChainStackItem()`).

(Note that my head is still spinning and I'm still not 100% sure that that's equivalent, but I am pretty sure it is.)

File: Python/errors.c [hunk 5]
```diff
@@ -586,7 +633,18 @@ _PyErr_ChainStackItem(_PyErr_StackItem *exc_info)
     exc2 = exc_info->exc_type;
```
Comment (gvanrossum) [inline Python/errors.c]: While you're here could you change 'exc' and 'exc2' to 'type' and 'type2' (or 'typ' and 'typ2')? It drives me nuts that we mix "exc[exption]" and "typ[e]" for the same concept (this must date back to Python 1 when the value was not an exception instance and the exception could be any object).

**Review by gvanrossum — COMMENTED** [review]
Okay, I'm through now. (Sorry for the premature send on the first batch.)

File: Python/errors.c (hunk 2 above)
Comment (iritkatriel) [reply Python/errors.c:81]: Can't hurt.

Looks like I can remove the assertion in the end though, because exc_info is only set to `exc_info->previous_item` if the latter is not NULL.   The old code would have crashed if it's NULL in the beginning.

`tstate->exc_info` is initialised to non-NULL here https://github.com/python/cpython/blob/24c10d2943c482c4d3ecc71d45df2d8c10fa5bb1/Python/pystate.c#L668
and I think it is always maintained non-NULL. But I will still add the assertion to make this expectation explicit.

File: Python/errors.c (hunk 3 above)
Comment (iritkatriel) [reply Python/errors.c:481]: Yes, this PR is just about exc_type. I won't be able to put such assertions for the tb.

**Review by gvanrossum — APPROVED** [review]
All good! (Alas, I'm still stuck trying to wrap my head around the `except*` implementation, so that one will probably have to wait until next week.)
